src/middlewared/middlewared/plugins/fc/fc.py

```diff
@@ -0,0 +1,142 @@
+import os.path
+import re
+import subprocess
+from functools import cache
+from pathlib import Path
+
+from middlewared.service import Service, filterable
+from middlewared.service_exception import CallError
+from middlewared.utils import filter_list
+from .utils import dmi_pci_slot_info
+
+FIBRE_CHANNEL_SEARCH_STRING = 'Fibre Channel'
+QLA2XXX_KERNEL_MODULE = 'qla2xxx_scst'
+FIBRE_CHANNEL_VENDOR = 'QLogic'
+FC_HOST_PAT = re.compile('/sys/devices/(.*)/(?P<host>host.*)/fc_host/(?P=host)')
+
+
+class FCService(Service):
+
+    class Config:
+        private = True
+
+    async def capable(self):
+        """
+        Returns True if the system is licensed for FIBRECHANNEL and contains
+        one or more Fibre Channel cards.  False otherwise.
+        """
+        if await self.middleware.call('system.is_enterprise'):
+            if await self.middleware.call('system.feature_enabled', 'FIBRECHANNEL'):
+                if await self.middleware.call('fc.hba_present'):
+                    return True
+        return False
+
+    @cache
+    def hba_present(self):
+        """
+        Return True if a Qlogic Fibre Channel card is present.  False otherwise.
+        """
+        lspci_cmd = ['lspci']
+        ret = subprocess.run(lspci_cmd, capture_output=True)
+        if ret.returncode:
+            error = ret.stderr.decode() if ret.stderr else ret.stdout.decode()
+            if not error:
+                error = 'No error message reported'
+            self.logger.debug('Failed to execute command: lspci with error: %r', error)
+            raise CallError(f'Failed to determine serial number/product: {error}')
+        for line in ret.stdout.decode().split('\n'):
+            if FIBRE_CHANNEL_SEARCH_STRING not in line:
+                continue
+            if FIBRE_CHANNEL_VENDOR not in line:
+                continue
+            # We matched!
+            return True
+        return False
+
+    @filterable
+    def fc_hosts(self, filters, options):
```

Comment:
Should we have `api_method` with `QueryArgs`?

Reply:
Yes.